rbac/providers/azure/aad_auth.py

```diff
@@ -87,3 +90,22 @@ def check_token(self, AUTH_TYPE):
                 "Accept": "application/json",
                 "Host": "graph.microsoft.com",
             }
+
+    def _check_token(self, AUTH_TYPE):
+        """check if Token exists and calls for then caches as a global variable 
+        if it does not"""
+        if self.graph_token is None or not self._time_left():
+            response = self.get_token(AUTH_TYPE)
+            self.token_creation_timestamp = dt.datetime.now()
+            self.graph_token = response.json()["access_token"]
+        if AUTH_TYPE.upper() == "SECRET":
+            return {"Authorization": self.graph_token}
+        elif AUTH_TYPE.upper() == "CERT":
+            return {"Authorization": self.graph_token, "Host": "graph.microsoft.com"}
+
+    def check_token_POST(self, AUTH_TYPE):
```

Comment:
This should be rearranged into:

- Refresh the token(token)
  - Check if token exists. Regenerate if not.
    - If so, check token expiration. Regenerate if expired.

- Build headers(token):
  - add AUTH_TYPE header
  - add Content-Type header

This makes it easier to find and modify the headers as necessary. Possibly to move and share them with other requests. It also makes it possible to refresh the token without returning headers (separating request preparation from request execution)

Reply:
_check_token refreshes the token, do we want to change that? 

how's this?
```
 def _check_token(self, AUTH_TYPE):
        """check if Token exists and calls for then caches as a global variable 
        if it does not"""
        if self.graph_token is None or not self._time_left():
            response = self.get_token(AUTH_TYPE)
            self.token_creation_timestamp = dt.datetime.now()
            self.graph_token = response.json()["access_token"]


    def _common_headers(self, AUTH_TYPE):
        if AUTH_TYPE.upper() == "SECRET":
            return {"Authorization": self.graph_token}
        elif AUTH_TYPE.upper() == "CERT":
            return {"Authorization": self.graph_token, "Host": "graph.microsoft.com"}


    def check_token_GET(self, AUTH_TYPE):
        """Check it Token exists and calls for and caches as global variable if it does not."""
        self._check_token(AUTH_TYPE)
        headers = self._common_headers(AUTH_TYPE)
        headers["Accept"] = "application/json"
        return headers


    def check_token_POST(self, AUTH_TYPE):
        """check if Token exists and calls for then caches as a global variable 
        if it does not"""
        self._check_token(AUTH_TYPE)
        headers = self._common_headers(AUTH_TYPE)
        headers["Content-Type"] = "application/json"
        return headers
```

Reply:
_check_token could be refresh_token(), load_valid_token(), or get_valid_token()... also the logic should return the token rather than mutate the object passed in. That is, pass in the graph_token and auth_type and return the (existing or new) token. You can do that without referencing 'self' (ie self could be removed from get_token() making get_token a standalone/static function (as it should be)

_common_headers() could be header_by_auth_type()
Regarding the final two: Does the token make a difference regarding GET/POST? I don't think so. Rather, the difference is in the I/O media type header (Accept or content_type). Thus the duplicate lines can be moved into a shared function, and the difference made more explicit.